mempool/v1/tx.go

```diff
@@ -0,0 +1,200 @@
+package v1
+
+import (
+	"time"
+
+	"github.com/tendermint/tendermint/internal/libs/clist"
+	tmsync "github.com/tendermint/tendermint/internal/libs/sync"
+	"github.com/tendermint/tendermint/mempool"
+	"github.com/tendermint/tendermint/types"
+)
+
+// WrappedTx defines a wrapper around a raw transaction with additional metadata
+// that is used for indexing.
+type WrappedTx struct {
+	// tx represents the raw binary transaction data
+	tx types.Tx
+
+	// hash defines the transaction hash and the primary key used in the mempool
+	hash [mempool.TxKeySize]byte
+
+	// height defines the height at which the transaction was validated at
+	height int64
+
+	// gasWanted defines the amount of gas the transaction sender requires
+	gasWanted int64
+
+	// priority defines the transaction's priority as specified by the application
+	// in the ResponseCheckTx response.
+	priority int64
+
+	// sender defines the transaction's sender as specified by the application in
+	// the ResponseCheckTx response.
+	sender string
+
+	// timestamp is the time at which the node first received the transaction from
+	// a peer. It is used as a second dimension is prioritizing transactions when
+	// two transactions have the same priority.
+	timestamp time.Time
+
+	// peers records a mapping of all peers that sent a given transaction
+	peers map[uint16]struct{}
+
+	// heapIndex defines the index of the item in the heap
+	heapIndex int
+
+	// gossipEl references the linked-list element in the gossip index
+	gossipEl *clist.CElement
+
+	// removed marks the transaction as removed from the mempool. This is set
+	// during RemoveTx and is needed due to the fact that a given existing
+	// transaction in the mempool can be evicted when it is simultaneously having
+	// a reCheckTx callback executed.
+	removed bool
+}
+
+func (wtx *WrappedTx) Size() int {
+	return len(wtx.tx)
+}
+
+// TxStore implements a thread-safe mapping of valid transaction(s).
+//
+// NOTE:
+// - Concurrent read-only access to a *WrappedTx object is OK. However, mutative
+//   access is not allowed. Regardless, it is not expected for the mempool to
+//   need mutative access.
+type TxStore struct {
+	mtx       tmsync.RWMutex
+	hashTxs   map[[mempool.TxKeySize]byte]*WrappedTx // primary index
+	senderTxs map[string]*WrappedTx                  // sender is defined by the ABCI application
+}
+
+func NewTxStore() *TxStore {
+	return &TxStore{
+		senderTxs: make(map[string]*WrappedTx),
+		hashTxs:   make(map[[mempool.TxKeySize]byte]*WrappedTx),
+	}
+}
+
+// Size returns the total number of transactions in the store.
+func (txs *TxStore) Size() int {
+	txs.mtx.RLock()
+	defer txs.mtx.RUnlock()
+
+	return len(txs.hashTxs)
+}
+
+// GetAllTxs returns all the transactions currently in the store.
+func (txs *TxStore) GetAllTxs() []*WrappedTx {
+	txs.mtx.RLock()
+	defer txs.mtx.RUnlock()
+
+	wTxs := make([]*WrappedTx, len(txs.hashTxs))
+	i := 0
+	for _, wtx := range txs.hashTxs {
+		wTxs[i] = wtx
```

Comment:
this always gives me the willies, because I've been bitten by [this bug](https://play.golang.org/p/jIDdZXpgsUJ) a couple of times. This code is safe, but you're one very innocent pointer twiddle away from really weird behavior.

Reply:
(the punchline is that `range` is implemented in a way that avoids realloc'ing for every iteration,)

Reply:
Hmm, I'm having trouble spotting the issue...

We're not referencing a non-pointer type as the value in the map are already references/pointers. Mutating your [example](https://play.golang.org/p/ZY7XQWT4LCN) gives us what we expect.